spec.md

```diff
@@ -0,0 +1,1813 @@
+  * _Assigned types_. Write `T($x)` (math. notation $`T_m(x)`$) for the type that `m` assigns to `$x`.
+    * _Special case: assigned kinds_. Note that `T($x)` may be `Ent`, `Rel`, `Att`, `Itf` (`Rol`), or `Val` (for value types) when `$x` is assigned a type as a concept — we speak of `T($x)` as the **type kind** of `m($x)` in this case.
+
+**Pattern satisfaction and answers**
+
+* _Pattern satisfaction_. A cmap `m` may **satisfy** a pattern `P`: 
+  > Intuitively, this means substituting the variables in `P` with the concepts assigned in `m` yields statements that are true in our type system.
+  
+  * _Definition_. Satisfication has two requirements: 
+    * **concept satisfaction** ("concepts assigned by `m` must conform with pattern `P`)
+    * **type satisfication** ("type assigned by `m` must conform with pattern `P`")
+
+    We define concept satisfaction in the _next section_, and only spell out _type satisfaction_ for now:
+    1. For any var `$x` we require $`m(x) : T_m(x)`$ to be true in the type system
+    1. If `$x isa $A` in `P` then require $`T_m(x) \leq m(A)`$
+    1. If `$x links ($B: $y)` in `P` then require $`T_m(y) < m(B)`$ and $`T_m(x) : \mathbf{Rel}(m(B))`$
+    1. If `$x links ($B[]: $y)` in `P` then require $`T_m(y) < m(B)`$ and $`T_m(x) : \mathbf{Rel}([m(B)])`$
+    1. If `$x has $B $y` in `P` then require $`T_m(x) < O_{m(B)}`$ and
+        * either $`T_m(y) < m(B)`$
+        * or $`T_m(y) = V$ for $V : \mathbf{Val}`$ and $`m(B) <_! V`$
+    1. If `$x has $B[] $y` in `P` then require $`T_m(x) < O_{m(B)}`$ and
+        * either $`T_m(y) < [m(B)]`$
+        * or $`T_m(y) = [V]`$ for $`V : \mathbf{Val}`$ and $`m(B) <_! V`$
+    1. If `$x in $y` in `P` then require $`[T_m(x)] \leq T_m(y)`$
+    1. If `$x = <EXPR>` in `P`, then require $`T(\mathrm{expr}) \leq T_m(x)`$ where $`T(\mathrm{expr})`$ is the type of the expression 
+        * _Note_: types of expressions can be computed recursively since assignments are acyclic.
+    1. If `$x = fun(<VARS>)` or `$x in fun(<VARS>)` in `P`, then require $`T(\mathrm{fun}) \leq T_m(y)`$ where $`T(\mathrm{fun})`$ is the output type of the function 
+  
+    _Remark_ 
+      * In the last to cases, we can replace $\leq$ with $=$ to compute the **minimal type assignment** (see "Answers" below).
+      * For **tvar**s `$x` we also pick $`m(x) : T_m(x)`$ as minimal as possible by default (e.g. `person : Ent` instead of `person : Type`).
+      * The extra cases for `has` are introduced to facilate working with computed values (of potentially non-attribute type) to match attributes.
+
+   * _Replacing **var**s with concepts_. When discussing pattern satisfaction, we always consider **fully variablized** statements (e.g. `$x isa $X`, `$X sub $Y`). This also determines satisfaction of **partially assigned** versions of these statements (e.g. `$x isa A`, `$X sub A`, `A sub $Y`, or `x isa $A`).
```

Comment:
I don't think the case `x isa $A` is a valid construction

Reply:
It's not a valid TypeQL query indeed, but it's a case we need to consider in our type system to describe pattern satisfaction later on!